### PR TITLE
[WIP] SWM

### PR DESCRIPTION
**Pull request by jasmainak**
@srcole are you okay if I copy the SWM code from neurodsp with proper attribution of course?

Just trying it on our simulated data to see how it works ...

### CODE REVIEW

**Comment by srcole** [comment]
Yup, fine by me. Let me know if you have any questions!

File: alphacsc/swm.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
+"""
+Code adopted from Voytek Lab package neurodsp:
+https://github.com/voytekresearch/neurodsp/blob/master/neurodsp/shape/swm.py
+
+The sliding window matching algorithm identifies the waveform shape of
+neural oscillations using correlations.
+"""
+
+# Authors: Scott Cole
+#          Mainak Jas <mainak.jas@telecom-paristech.fr>
```
Comment (agramfort) [inline alphacsc/swm.py]: license?

Comment (srcole) [inline alphacsc/swm.py]: the neurodsp module is under the MIT license, but if you want to use a different license, I'm fine with that

Comment (jasmainak) [inline alphacsc/swm.py]: We'll use BSD. I guess they are pretty much the same :)

**Comment by agramfort** [comment]
ok for DSP

File: alphacsc/swm.py [hunk 2]
```diff
@@ -136,14 +137,10 @@ def _compute_J(x, window_starts, L):
...
-    for i in range(N_windows):
-        for j in range(1, N_windows):
-            window_diff = windows[i] - windows[j]
-            d_temp = np.sum(window_diff**2) / float(L)
-            d.append(d_temp)
-    # Calculate cost, the average difference, roughly
-    J = np.sum(d) / float(2 * (N_windows - 1))
+    dist = pdist(np.vstack(windows),
+                 lambda u, v: np.sum((u - v) ** 2))
+    J = np.sum(dist) / float(L * (N_windows - 1))
```
Comment (agramfort) [inline alphacsc/swm.py]: can't you use http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.euclidean_distances.html ?

Comment (jasmainak) [inline alphacsc/swm.py]: what's the difference? I'm using `pdist` from scipy now

**Comment by agramfort** [comment]
time it :)

**Comment by jasmainak** [comment]
> time it :)

haha okay.

So, I had to decrease the temperature `T` to make the cost function go down in our simulation. I think equation (1) was missing a factor of 2, as far as I can tell. Probably doesn't matter in terms of final result. The zscoring trick seems to be problematic when you don't have much noise. The noise gets amplified in time periods with no atoms.

So far, I still don't get atoms (or windows) that look sensible. As I understand, @srcole has not implemented the parallel tempering part of the algorithm. Not sure how much it matters ...

**Comment by srcole** [comment]
> I think equation (1) was missing a factor of 2, as far as I can tell.
Yeah, I see that now. Don't see how it could make a difference.

> So far, I still don't get atoms (or windows) that look sensible. 
What is your simulated data like?

> As I understand, @srcole has not implemented the parallel tempering part of the algorithm. Not sure how much it matters ...
Right. I didn't include that for my purposes (I was getting reasonable looking results for M1 ECoG data). But from the text of the paper, it sounds like it could be important in some scenarios.

**Comment by jasmainak** [comment]
> What is your simulated data like?

Good question. I just realized what was breaking the algorithm. It's the fact that the simulated data has atoms (recurring patterns) of varying amplitude. It's the [same with MoTIF](https://alphacsc.github.io/auto_examples/plot_simulate_motif.html#sphx-glr-auto-examples-plot-simulate-motif-py). It seems to work much better if it's constant amplitude.

**Comment by agramfort** [comment]
@srcole did you share some of these M1 ECoG data ? if so we could try the alphacsc code to see how it looks like? how does that sound @srcole ?

**Comment by srcole** [comment]
sounds good! Here is one used for our lab tutorials: https://github.com/voytekresearch/tutorials/blob/master/exampledata.npy

**Comment by agramfort** [comment]
great

can we make an alphacsc example from it?

what paper do we cite in the example for origin of the data?

**Comment by srcole** [comment]
Yeah, you can use it.

It was Cora de Hemptinne and Phil Starr who collected the data for their paper here:
> de Hemptinne, C. et al., 2015. Therapeutic deep brain stimulation reduces cortical phase-amplitude
coupling in Parkinson’s disease. Nature Neuroscience, 18(5), pp.779–786. Available at:
http://dx.doi.org/10.1038/nn.3997.

And they gave me permission in our paper to release the [full data set](https://github.com/voytekresearch/Cole_2017)
> Cole SR. et al., 2017. Nonsinusoidal beta oscillations reflect cortical pathophysiology in Parkinson’s disease. Nature Neuroscience, 18(5), pp.779–786. Available at:
http://dx.doi.org/10.1038/nn.3997.

**Comment by jasmainak** [comment]
okay, merged into master for now. We iterate further from [here](https://alphacsc.github.io/auto_examples/plot_simulate_swm.html#sphx-glr-auto-examples-plot-simulate-swm-py). Thanks @agramfort and @srcole for the valuable feedback

**Pull request merged**